### ISSUE TITLE
Issue with blocked connection

### ISSUE THREAD

**Issue by ono** [rabbitmq/rabbitmq-erlang-client#93]
We had an issue with blocked connections which could cause the serious damage to your system.

#### Step 1. create a test queue on Rabbit MQ

Let's name it `test`.

#### Step 2. simulate the blocked connection by setting lower memory threshold

For example, set the following value in your `rabbitmq.config`.

```erlang
[
    {rabbit, [{vm_memory_high_watermark, {absolute, "10MiB"}}]}
].
```

Then restart RabbitMQ and try making a connection.
Then you will see the [connection state is blocked](https://www.dropbox.com/s/2cts633jo6dru2j/za0a.png?dl=0) when you check the connection on admin panel.

#### Step 3. try publish a message to the server

If you call the following code, it will never return.

```erlang
test_blocked() ->
    {ok, Connection} =
        amqp_connection:start(#amqp_params_network{host = "localhost"}),
    {ok, Channel} = amqp_connection:open_channel(Connection),

    % Problem 1: I'd expect `call/3` to return `blocking` instead of `ok` but it returns `ok`
    ok = amqp_channel:call(Channel,
                           #'basic.publish'{
                               exchange = <<"">>,
                               routing_key = <<"test">>
                           },
                           #amqp_msg{payload = <<"Hello World!">>}),

    % Problem 2: The program will hang here. It will never return any value or raise an exception.
    ok = amqp_channel:close(Channel),
    ok.
```

#### Environment

I can reproduce the issue at least with:

- OTP: 20
- amqp_client and rabbit_common: 3.6.12
- RabbitMQ server 3.6.6

**Comment by michaelklishin** [comment]
This works [as expected](http://www.rabbitmq.com/alarms.html). A blank node usually uses 70-80 MB of RAM on a 64 bit machine, so it immediately goes into the alarm and any connection that publishes will be blocked. There are very visible entries in the log about this.

**Comment by michaelklishin** [comment]
And your connection state is not blocked, it is "blocking", which means as soon as you publish, it will be blocked. You could consume on that connection just fine.

**Comment by michaelklishin** [comment]
Clients can [register notifications](http://www.rabbitmq.com/connection-blocked.html) for when a connection is blocked or unblocked. If you have further questions, please take them to rabbitmq-users, our mailing list.

**Comment by ono** [comment]
Thanks for the response but i wonder if the issue was understood correctly.

The problem I'm reporting is that it does NOT block the publish. `basic.publish` accepts the call and returns `ok` instead of `blocking`. And `amqp_channel:close/1` never return any state - the program just stops there.

Have you tried to run the sample code with the steps I provided? I will be very surprised if that's intended behaviour.

**Comment by michaelklishin** [comment]
Thank you for your time.

Team RabbitMQ uses GitHub issues for specific actionable items engineers can work on. This assumes two things:

1. GitHub issues are not used for questions, investigations, root cause analysis, discussions of potential issues, etc (as defined by this team)
2. We have a certain amount of information to work with

We get at least a dozen of questions through various venues every single day, often quite light on details.
At that rate GitHub issues can very quickly turn into a something impossible to navigate and make sense of even for our team. Because of that questions, investigations, root cause analysis, discussions of potential features are all considered to be mailing list material by our team. Please post this to [rabbitmq-users](https://groups.google.com/forum/#!forum/rabbitmq-users).

Getting all the details necessary to reproduce an issue, make a conclusion or even form a hypothesis about what's happening can take a fair amount of time. Our team is multiple orders of magnitude smaller than the RabbitMQ community. Please help others help you by providing a way to reproduce the behavior you're
observing, or at least sharing as much relevant information as possible [on the list](https://groups.google.com/forum/#!forum/rabbitmq-users):

 * Server, client library and plugin (if applicable) versions used
 * Server logs
 * A code example or terminal transcript that can be used to reproduce
 * **Full** exception stack traces (not a single line message)
 * `rabbitmqctl status` (and, if possible, `rabbitmqctl environment` output)
 * Other relevant things about the environment and workload, e.g. [a traffic capture](https://www.rabbitmq.com/amqp-wireshark.html)

Feel free to edit out hostnames and other potentially sensitive information.

When/if we have enough details and evidence we'd be happy to file a new issue.

Thank you.

**Comment by michaelklishin** [comment]
@ono `basic.publish` is an entirely asynchronous operation. The client has no way of knowing what happened to the message besides [using publisher confirms](http://www.rabbitmq.com/confirms.html). When an alarm is in effect, connections transition to the `blocking` state I've mentioned before. This is because we do not want to block consumers. As soon as a `basic.publish`, a content header (message properties) or a content body frame is seen on a connection, it will be blocked and no further TCP socket reads will happen. Again, the client has no way of knowing about that since that's how TCP works and that's why `connection.[un]blocked` protocol extensions were introduced. You can register a handler and see for yourself, or try to publish more messages and observe message rates.
